server/src/main/java/org/opensearch/index/LogByteSizeMergePolicyProvider.java

```diff
@@ -0,0 +1,166 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ *
+ * The OpenSearch Contributors require contributions made to
+ * this file be licensed under the Apache-2.0 license or a
+ * compatible open source license.
+ */
+
+package org.opensearch.index;
+
+import org.apache.logging.log4j.Logger;
+import org.apache.lucene.index.LogByteSizeMergePolicy;
+import org.apache.lucene.index.MergePolicy;
+import org.apache.lucene.index.NoMergePolicy;
+import org.opensearch.common.settings.Setting;
+import org.opensearch.core.common.unit.ByteSizeUnit;
+import org.opensearch.core.common.unit.ByteSizeValue;
+
+import static org.apache.lucene.index.LogMergePolicy.DEFAULT_MAX_MERGE_DOCS;
+import static org.apache.lucene.index.LogMergePolicy.DEFAULT_NO_CFS_RATIO;
+
+/**
+ * <p>
+ * The LogByteSizeMergePolicy is an alternative merge policy primarily used here to optimize the merging of segments in scenarios
+ * with index with timestamps.
+ * While the TieredMergePolicy is the default choice, the LogByteSizeMergePolicy can be configured
+ * as the default merge policy for time-index data using the <code>index.datastream_merge.policy</code> setting.
+ *
+ * <p>
+ * Unlike the TieredMergePolicy, which prioritizes merging segments of equal sizes, the LogByteSizeMergePolicy
+ * specializes in merging adjacent segments efficiently.
+ * This characteristic makes it particularly well-suited for range queries on time-index data.
+ * Typically, adjacent segments in time-index data often contain documents with similar timestamps.
+ * When these segments are merged, the resulting segment covers a range of timestamps with reduced overlap compared
+ * to the adjacent segments. This reduced overlap remains even as segments grow older and larger,
+ * which can significantly benefit range queries on timestamps.
+ *
+ * <p>
+ * In contrast, the TieredMergePolicy does not honor this timestamp range optimization. It focuses on merging segments
+ * of equal sizes and does not consider adjacency. Consequently, as segments grow older and larger,
+ * the overlap of timestamp ranges among adjacent segments managed by TieredMergePolicy can increase.
+ * This can lead to inefficiencies in range queries on timestamps, as the number of segments to be scanned
+ * within a given timestamp range could become high.
+ *
+ * @opensearch.internal
+ */
+public class LogByteSizeMergePolicyProvider implements MergePolicyProvider {
```

Comment:
Since we use `MergePolicyProvider` now, could we make this class (and TieredXxx counterpart) package private?

```suggestion
class LogByteSizeMergePolicyProvider implements MergePolicyProvider {
```

Reply:
both of them are used in `org.opensearch.common.settings.IndexScopedSettings`, so it cannot be package-private.